src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs

```diff
@@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
         ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
         public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
         {
+            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;
+
+            if (argumentList != null)
+            {
+                switch (argumentList.Count)
+                {
+                    case 0:
+                        return Call(instance, method);
+                    case 1:
+                        return Call(instance, method, argumentList[0]);
+                    case 2:
+                        return Call(instance, method, argumentList[0], argumentList[1]);
+                    case 3:
+                        return Call(instance, method, argumentList[0], argumentList[1], argumentList[2]);
+                }
+
+                if (instance == null)
+                {
+                    switch (argumentList.Count)
```

Comment:
Fair point; I'll tweak this.

Sent from my Windows Phone
________________________________
From: Stephen Toub<mailto:notifications@github.com>
Sent: ‎9/‎17/‎2015 4:00 PM
To: dotnet/corefx<mailto:corefx@noreply.github.com>
Cc: Bart De Smet<mailto:bartde@microsoft.com>
Subject: Re: [corefx] Use optimized N-ary MethodCallExpression nodes from all Call factories (#3288)


In src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3288%23discussion_r39809201&data=01%7c01%7cbartde%40microsoft.com%7c93f7a5e8c2df4bd8a8cd08d2bfb3c825%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=nvG62BW54JPueHZU9Z5g7V3mqVwYKrxVsjFXe6ic20Y%3d>:

> +            {
> +                switch (argumentList.Count)
> +                {
> +                    case 0:
> +                        return Call(instance, method);
> +                    case 1:
> +                        return Call(instance, method, argumentList[0]);
> +                    case 2:
> +                        return Call(instance, method, argumentList[0], argumentList[1]);
> +                    case 3:
> +                        return Call(instance, method, argumentList[0], argumentList[1], argumentList[2]);
> +                }
> +
> +                if (instance == null)
> +                {
> +                    switch (argumentList.Count)


Minor, but you could save the count into a local earlier such that in this case we don't need to make the extra interface call.

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3288%2ffiles%23r39809201&data=01%7c01%7cbartde%40microsoft.com%7c93f7a5e8c2df4bd8a8cd08d2bfb3c825%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=3g0vAW7i83N06zr1%2bklv29zOzDBBCaXXkZUZpoD7Ras%3d>.